### PR TITLE
:white_check_mark: Settle and on-chain balance assertion

### LINKED ISSUE #28: Test: Settle & On-Chain balance assertion
The CI test should settle the channel and check the on-chain balances afterwards.

### PR DESCRIPTION
**Pull request by cryptphil**
Closes #28

### CODE REVIEW

**Comment by ggwpez** [comment]
I dont think this will automatically close the issue, have a look at https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword for keywords that would close it.

File: cmd/demo/demo_test.go [hunk 1]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
 
 const blockTime = 5 * time.Second
 
+const sendIterations = 25
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: You could name it `numUpdates` since you were looking for a better name yesterday afaik.

File: cmd/demo/demo_test.go [hunk 2]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
+
+const ethUrl = "ws://127.0.0.1:8545"
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: If you have so many consts you can accumulate them with:  
```go
const (
    addressAlice = "0x05e71027e7d3bd6261de7634cf50F0e2142067C4"
    ethUrl = "ws://127.0.0.1:8545"
    …
)
```
This also gets rid of all the newlines.

File: cmd/demo/demo_test.go [hunk 3]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
-	// Start Bob
+	// Start Bob.
 	bob, err := nodeCmd("bob")
 	require.NoError(t, err)
 	require.NoError(t, bob.Start())
 	defer bob.Process.Kill()
-	time.Sleep(5 * time.Second) // give bob some time to initialize
+	time.Sleep(5 * time.Second) // Give Bob some time to initialize.
+
+	// Set the initial on-chain balances from Alice and Bob.
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: I dont wanna be nit-picky but you dont really "Set" them here but "Get" as the name `getOnChainBals`  already suggests. Feel free to ignore it though :laughing:

File: cmd/demo/demo_test.go [hunk 4]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
+	initBals, err := getOnChainBals()
+	require.NoError(t, err)
+	t.Logf("Initial on-chain balances: Alice = %f, Bob = %f", initBals[0], initBals[1])
 
-	// Alice opens channel with Bob
+	// Alice opens channel with Bob.
 	require.NoError(t, alice.sendCommand("open bob 1000 1000\n"), "proposing channel")
+	time.Sleep(3 * time.Second) // Ensure that Bob really received the proposal.
```
Comment (ggwpez) [inline cmd/demo/demo_test.go:86]: You can shorten this to 1s, it should actually take more like 1ms.

File: cmd/demo/demo_test.go [hunk 5]
```diff
@@ -6,8 +6,14 @@
 package demo_test
 
 import (
+	"context"
 	"fmt"
+	"github.com/ethereum/go-ethereum/common"
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: I dont know if your IDE supports this, but we use `goimports` to keep the imports ordered. In this case the `go-ethereum` imports should be between `"time"` and `"testify/require"` with one newline apart.  

Im sorry that there is no `CONTRIBUTING.md` otherwise this would be explained there…

File: cmd/demo/demo_test.go [hunk 6]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
 		require.NoError(t, bob.sendCommand("send alice 2\n"))
 		time.Sleep(100 * time.Millisecond)
 	}
+
+	t.Log("Closing channel…")
+	require.NoError(t, alice.sendCommand("close bob\n"))
+	// Wait 2 blocks for the settle and withdrawal transactions plus some additional seconds.
+	time.Sleep(2*blockTime + 5*time.Second)
+
+	// Set the final balances from Alice and Bob after the settlement.
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]:
```suggestion
	// Get the final balances from Alice and Bob after the settlement.
```
same as above.

File: cmd/demo/demo_test.go [hunk 7]
```diff
@@ -6,8 +6,14 @@
 package demo_test
```
Comment (ggwpez) [inline cmd/demo/demo_test.go:7]: Could you please update the Copyright year to 2021?

File: cmd/demo/demo_test.go [hunk 8]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
+	t.Logf("Final on-chain balances: Alice = %f, Bob = %f", finalBals[0], finalBals[1])
+
+	var diffBals [2]float64
+
+	// Calculate the differences between the initial and final balances.
+	diffBals[0], _ = finalBals[0].Sub(finalBals[0], initBals[0]).Float64()
+	diffBals[1], _ = finalBals[1].Sub(finalBals[1], initBals[1]).Float64()
+
+	// Check the on-chain balance differences while allowing a higher deviation for Alice.
+	require.Less(t, diffBals[0], sendIterations+0.1)
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: You can use `assert.InEpsilon(t, sendIterations, diffBals[0], 0.1)` here and below, then you need just one check.
This is the only place where it is actually an assert instead of a require so it is probably not obvious :see_no_evil: 

I also tried to set the gas price of the chain to `0`, which then produces the exact values of `+-25.0` but i think it is better to run it with gas like you do right now.

File: cmd/demo/demo_test.go [hunk 9]
```diff
@@ -46,40 +52,102 @@ func nodeCmd(name string) (*Cmd, error) {
...
 
 func (cmd *Cmd) sendCommand(str string) error {
 	_, err := io.WriteString(cmd.stdin, str)
 	return err
 }
+
+func getOnChainBals() ([2]*big.Float, error) {
+	var bals [2]*big.Float
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
```
Comment (ggwpez) [inline cmd/demo/demo_test.go]: Did you try it with lower timeout? 30s seems a bit excessive for just two queries. 3s should work too.  
You can also keep it and use `DialContext` below instead of `Dial` then it makes more sense AFAIK.

**Review by ggwpez — CHANGES_REQUESTED** [review]

File: cmd/demo/demo_test.go (hunk 4 above)
Comment (cryptphil) [reply cmd/demo/demo_test.go:86]: Tested it also with 1s, still, often Bob does not accept any proposal or rather just enters "y" without actually having received the proposal yet.

Comment (ggwpez) [reply cmd/demo/demo_test.go:86]: Ok then lets leave it as it, seems like a bug to me though :thinking:

**Review by ggwpez — APPROVED** [review]
